// piral-cli/lib/apps/add-piral-instance-pilet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPiralInstancePilet = exports.addPiralInstancePiletDefaults = void 0;
const path_1 = require("path");
const common_1 = require("../common");
const types_1 = require("../types");
exports.addPiralInstancePiletDefaults = {
    logLevel: types_1.LogLevels.info,
    app: undefined,
    source: '.',
    selected: false,
    npmClient: undefined,
};
function addPiralInstancePilet(baseDir = process.cwd(), options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { npmClient: defaultNpmClient = exports.addPiralInstancePiletDefaults.npmClient, logLevel = exports.addPiralInstancePiletDefaults.logLevel, source = exports.addPiralInstancePiletDefaults.source, selected = exports.addPiralInstancePiletDefaults.selected, app = exports.addPiralInstancePiletDefaults.app, } = options;
        const fullBase = (0, path_1.resolve)(process.cwd(), baseDir);
        (0, common_1.setLogLevel)(logLevel);
        (0, common_1.progress)('Reading configuration ...');
        const npmClient = yield (0, common_1.determineNpmClient)(fullBase, defaultNpmClient);
        const allEntries = yield (0, common_1.matchAnyPilet)(fullBase, [source]);
        const tasks = allEntries.map((entryModule) => __awaiter(this, void 0, void 0, function* () {
            const targetDir = (0, path_1.dirname)(entryModule);
            const piletJson = 'pilet.json';
            const piletJsonPath = yield (0, common_1.findFile)(targetDir, piletJson);
            if (piletJsonPath) {
                const piletJsonDir = (0, path_1.dirname)(piletJsonPath);
                const root = yield (0, common_1.findPiletRoot)(piletJsonDir);
                const oldContent = yield (0, common_1.readJson)(piletJsonDir, piletJson);
                const [appName] = yield (0, common_1.installPiralInstance)(app, fullBase, root, npmClient);
                const newContent = Object.assign(Object.assign({}, oldContent), { piralInstances: Object.assign(Object.assign({}, oldContent.piralInstances), { [appName]: {
                            selected,
                        } }) });
                yield (0, common_1.writeJson)(piletJsonDir, piletJson, newContent, true);
            }
            else {
                (0, common_1.log)('piletJsonNotAvailable_0180', targetDir);
            }
        }));
        yield Promise.all(tasks);
        (0, common_1.logDone)(`Added the Piral instance!`);
    });
}
exports.addPiralInstancePilet = addPiralInstancePilet;
