// piral-core/lib/components/ExtensionSlot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionSlot = void 0;
const React = require("react");
const piral_base_1 = require("piral-base");
const wrapComponent_1 = require("./wrapComponent");
const hooks_1 = require("../hooks");
const utils_1 = require("../utils");
const wrapper = ({ children }) => (0, utils_1.defaultRender)(children);
const renderExtensions = [
    {
        component: (props) => {
            const context = (0, hooks_1.useGlobalStateContext)();
            const converters = context.converters;
            const piral = context.apis._;
            const { component, props: args } = props.params;
            const Component = React.useMemo(() => (0, wrapComponent_1.wrapComponent)(converters, component, { piral }, wrapper), [component]);
            return React.createElement(Component, Object.assign({}, args));
        },
        defaults: {},
        pilet: '',
        reference: {
            displayName: 'AnyComponent',
        },
    },
];
function defaultOrder(extensions) {
    return extensions;
}
/**
 * The extension slot component to be used when the available
 * extensions of a given name should be rendered at a specific
 * location.
 */
function ExtensionSlot(props) {
    const { name, render = utils_1.defaultRender, empty, params, children, emptySkipsRender = false, order = defaultOrder, } = props;
    const extensions = (0, hooks_1.useGlobalState)((s) => (name ? s.registry.extensions[name] || utils_1.none : renderExtensions));
    const isEmpty = extensions.length === 0 && (0, piral_base_1.isfunc)(empty);
    const content = isEmpty
        ? [(0, utils_1.defaultRender)(empty(), 'empty')]
        : order(extensions).map(({ component: Component, reference, defaults = {} }, i) => (React.createElement(Component, { key: `${(reference === null || reference === void 0 ? void 0 : reference.displayName) || '_'}${i}`, children: children, params: Object.assign(Object.assign({}, defaults), params) })));
    if (isEmpty && emptySkipsRender) {
        return content[0];
    }
    return render(content);
}
exports.ExtensionSlot = ExtensionSlot;
ExtensionSlot.displayName = `ExtensionSlot`;
