// piral-cli/lib/injectors/pilet-injector.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
const path_1 = require("path");
const fs_1 = require("fs");
const log_1 = require("../common/log");
const spec_1 = require("../common/spec");
const config_1 = require("../common/config");
const external_1 = require("../external");
function fillPiletMeta(pilet, metaFile, subPath) {
    const { root, bundler } = pilet;
    const metaPath = (0, path_1.join)(root, metaFile);
    const packagePath = (0, path_1.join)(root, 'package.json');
    const def = external_1.jju.parse((0, fs_1.readFileSync)(packagePath, 'utf8'));
    const metaOverride = (0, fs_1.existsSync)(metaPath) ? external_1.jju.parse((0, fs_1.readFileSync)(metaPath, 'utf8')) : undefined;
    const file = bundler.bundle.name.replace(/^[\/\\]/, '');
    const target = (0, path_1.join)(bundler.bundle.dir, file);
    pilet.getMeta = (parentPath) => {
        const basePath = `${parentPath}${subPath}`;
        const url = new url_1.URL(file, basePath);
        return Object.assign(Object.assign(Object.assign({ custom: def.custom, config: def.piletConfig }, metaOverride), { name: def.name, version: def.version, link: `${url.href}?updated=${Date.now()}` }), (0, spec_1.getPiletSpecMeta)(target, basePath));
    };
}
function loadFeed(feed) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const response = yield external_1.axios.default.get(feed);
            if (Array.isArray(response.data)) {
                return response.data;
            }
            else if (Array.isArray((_a = response.data) === null || _a === void 0 ? void 0 : _a.items)) {
                return response.data.items;
            }
            else {
                return [response.data];
            }
        }
        catch (e) {
            (0, log_1.log)('generalWarning_0001', `Couldn't load feed at ${feed}.`);
        }
    });
}
class PiletInjector {
    constructor(config, serverConfig, core) {
        this.config = config;
        this.serverConfig = serverConfig;
        if (this.config.active) {
            const { pilets, api, publicUrl, assetUrl } = config;
            this.indexPath = `${publicUrl}index.html`;
            const cbs = {};
            core.on('user-connected', (e) => {
                const baseUrl = assetUrl || e.req.headers.origin;
                if (e.target === '*' && e.url === api.substring(1)) {
                    cbs[e.id] = {
                        baseUrl,
                        notify: (msg) => e.ws.send(msg),
                    };
                }
            });
            core.on('user-disconnected', (e) => {
                delete cbs[e.id];
            });
            pilets.forEach((p, i) => p.bundler.on(() => {
                fillPiletMeta(p, config.meta, `/${i}/`);
                for (const id of Object.keys(cbs)) {
                    const { baseUrl, notify } = cbs[id];
                    const meta = this.getPiletMeta(baseUrl, p);
                    notify(meta);
                }
            }));
        }
    }
    get active() {
        return this.config.active;
    }
    set active(value) {
        this.config.active = value;
    }
    get name() {
        return 'pilet-injector';
    }
    getOptions() {
        return {};
    }
    setOptions() { }
    getPiletApi(baseUrl) {
        const { api } = this.config;
        if (/^https?:/.test(api)) {
            return api;
        }
        else if (baseUrl) {
            return `${baseUrl}${api}`;
        }
        else {
            const { ssl, port } = this.serverConfig;
            const { host } = config_1.config;
            return `${ssl ? 'https' : 'http'}://${host}:${port}${api}`;
        }
    }
    getPiletMeta(baseUrl, pilet) {
        const basePath = this.getPiletApi(baseUrl);
        return JSON.stringify(pilet.getMeta(basePath));
    }
    getIndexMeta(baseUrl) {
        return __awaiter(this, void 0, void 0, function* () {
            const { pilets, feed } = this.config;
            const basePath = this.getPiletApi(baseUrl);
            const localPilets = pilets.map((pilet) => { var _a; return (_a = pilet.getMeta) === null || _a === void 0 ? void 0 : _a.call(pilet, basePath); }).filter(Boolean);
            const mergedPilets = this.mergePilets(localPilets, yield this.loadRemoteFeed(feed));
            return JSON.stringify(mergedPilets);
        });
    }
    loadRemoteFeed(feed) {
        return __awaiter(this, void 0, void 0, function* () {
            if (feed) {
                const feeds = Array.isArray(feed) ? feed : [feed];
                return yield Promise.all(feeds.map(loadFeed));
            }
        });
    }
    mergePilets(localPilets, remoteFeeds) {
        if (!remoteFeeds || !Array.isArray(remoteFeeds)) {
            return localPilets;
        }
        const { mergeConfig = false } = this.config;
        const names = localPilets.map((pilet) => pilet.name);
        const merged = [...localPilets];
        for (const remotePilets of remoteFeeds) {
            if (!Array.isArray(remotePilets)) {
                continue;
            }
            const newPilets = remotePilets.filter((pilet) => {
                if (!pilet || typeof pilet !== 'object') {
                    return false;
                }
                const name = pilet.name;
                const isNew = name !== undefined && !names.includes(name);
                if (!isNew && mergeConfig) {
                    const existing = merged.find(m => m.name === name);
                    if (existing.config === undefined) {
                        existing.config = pilet.config;
                    }
                    else if (pilet.config !== undefined) {
                        Object.assign(existing.config, pilet.config);
                    }
                }
                return isNew;
            });
            names.push(...newPilets.map((p) => p.name));
            merged.push(...newPilets);
        }
        return merged;
    }
    sendContent(content, type, url) {
        const { headers } = this.config;
        return {
            injector: { name: this.name },
            headers: Object.assign(Object.assign({}, headers), { 'content-type': type, 'cache-control': 'no-cache, no-store, must-revalidate', pragma: 'no-cache', expires: '0' }),
            status: { code: 200 },
            url,
            content,
        };
    }
    sendFile(target, url) {
        var _a;
        const content = (0, fs_1.readFileSync)(target);
        const type = (_a = external_1.mime.getType(target)) !== null && _a !== void 0 ? _a : 'application/octet-stream';
        return this.sendContent(content, type, url);
    }
    sendResponse(path, url, baseUrl) {
        return __awaiter(this, void 0, void 0, function* () {
            const { pilets } = this.config;
            const [index, ...rest] = path.split('/');
            const pilet = pilets[+index];
            const bundler = pilet === null || pilet === void 0 ? void 0 : pilet.bundler;
            if (!path) {
                yield (bundler === null || bundler === void 0 ? void 0 : bundler.ready());
                const content = yield this.getIndexMeta(baseUrl);
                return this.sendContent(content, 'application/json', url);
            }
            else {
                return bundler === null || bundler === void 0 ? void 0 : bundler.ready().then(() => {
                    const target = (0, path_1.join)(bundler.bundle.dir, rest.join('/'));
                    if ((0, fs_1.existsSync)(target) && (0, fs_1.statSync)(target).isFile()) {
                        return this.sendFile(target, url);
                    }
                });
            }
        });
    }
    sendIndexFile(target, url, baseUrl) {
        const indexHtml = (0, fs_1.readFileSync)(target, 'utf8');
        // mechanism to inject server side debug piletApi config into piral emulator
        const windowInjectionScript = `window['dbg:pilet-api'] = '${this.getPiletApi(baseUrl)}';`;
        const findStr = `<script`;
        const replaceStr = `<script>/* Pilet Debugging Emulator Config Injection */${windowInjectionScript}</script><script`;
        const content = indexHtml.replace(`${findStr}`, `${replaceStr}`);
        return this.sendContent(content, external_1.mime.getType(target), url);
    }
    handle(req) {
        const { app, api, publicUrl, assetUrl } = this.config;
        const baseUrl = assetUrl || (req.headers.host ? `${req.encrypted ? 'https' : 'http'}://${req.headers.host}` : undefined);
        if (!req.target) {
            if (req.url.startsWith(publicUrl)) {
                const path = req.url.substring(publicUrl.length).split('?')[0];
                const target = (0, path_1.join)(app, path);
                if ((0, fs_1.existsSync)(target) && (0, fs_1.statSync)(target).isFile()) {
                    if (req.url === this.indexPath) {
                        return this.sendIndexFile(target, req.url, baseUrl);
                    }
                    else {
                        return this.sendFile(target, req.url);
                    }
                }
                else if (req.url !== this.indexPath) {
                    return this.handle(Object.assign(Object.assign({}, req), { url: this.indexPath }));
                }
            }
            return undefined;
        }
        else if (req.target === api) {
            const path = req.url.substring(1).split('?')[0];
            return this.sendResponse(path, req.url, baseUrl);
        }
    }
}
exports.default = PiletInjector;
