// piral-cli/lib/injectors/piral-injector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_1 = require("fs");
const external_1 = require("../external");
/**
 * The maximum amount of retries when sending a response
 */
const maxRetrySendResponse = 4;
class PiralInjector {
    constructor(options, _config, core) {
        this.config = options;
        if (this.config.active) {
            const api = '/$events';
            const cbs = {};
            core.on('user-connected', (e) => {
                if (e.target === '*' && e.url === api.substring(1)) {
                    cbs[e.id] = (msg) => e.ws.send(msg);
                }
            });
            core.on('user-disconnected', (e) => {
                delete cbs[e.id];
            });
            this.config.bundler.on((args) => {
                for (const id of Object.keys(cbs)) {
                    cbs[id](JSON.stringify(args));
                }
            });
        }
    }
    get active() {
        return this.config.active;
    }
    set active(value) {
        this.config.active = value;
    }
    get name() {
        return 'piral-injector';
    }
    getOptions() {
        return {};
    }
    setOptions() { }
    sendContent(content, type, url) {
        const { headers } = this.config;
        return {
            injector: { name: this.name },
            headers: Object.assign(Object.assign({}, headers), { 'content-type': type, 'cache-control': 'no-cache, no-store, must-revalidate', pragma: 'no-cache', expires: '0' }),
            status: { code: 200 },
            url,
            content,
        };
    }
    sendIndexFile(target, url) {
        const indexHtml = (0, fs_1.readFileSync)(target, 'utf8');
        const { feed } = this.config;
        if (feed) {
            // mechanism to inject server side debug piletApi config into piral emulator
            const windowInjectionScript = `window['dbg:pilet-api'] = '${feed}';`;
            const findStr = `<script`;
            const replaceStr = `<script>/* Pilet Debugging Emulator Config Injection */${windowInjectionScript}</script><script`;
            const content = indexHtml.replace(`${findStr}`, `${replaceStr}`);
            return this.sendContent(content, external_1.mime.getType(target), url);
        }
        return this.sendContent(indexHtml, external_1.mime.getType(target), url);
    }
    sendResponse(path, target, dir, url, recursionDepth = 0) {
        var _a;
        if (recursionDepth > maxRetrySendResponse) {
            return undefined;
        }
        const { bundler } = this.config;
        const newTarget = (0, path_1.join)(bundler.bundle.dir, bundler.bundle.name);
        if (!path || !(0, fs_1.existsSync)(target) || !(0, fs_1.statSync)(target).isFile()) {
            return this.sendResponse(bundler.bundle.name, newTarget, dir, url, recursionDepth + 1);
        }
        else if (target === newTarget) {
            return this.sendIndexFile(target, url);
        }
        const type = (_a = external_1.mime.getType(target)) !== null && _a !== void 0 ? _a : 'application/octet-stream';
        return this.sendContent((0, fs_1.readFileSync)(target), type, url);
    }
    handle(req) {
        if (!req.target) {
            const { bundler, publicUrl } = this.config;
            if (req.url.startsWith(publicUrl) || `${req.url}/` === publicUrl) {
                const pathLength = publicUrl.length || 1;
                const path = req.url.substring(pathLength);
                const dir = bundler.bundle.dir;
                const target = (0, path_1.join)(dir, path.split('?')[0]);
                return bundler.ready().then(() => this.sendResponse(path, target, dir, req.url));
            }
        }
    }
}
exports.default = PiralInjector;
