// piral-core/lib/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPiletOptions = void 0;
const piral_base_1 = require("piral-base");
const modules_1 = require("./modules");
const app_codegen_1 = require("../app.codegen");
function createPiletOptions({ hooks, context, loaders, loaderConfig, availablePilets, strategy, createApi, loadPilet, requestPilets, shareDependencies, debug, }) {
    const options = {
        config: loaderConfig,
        strategy,
        loadPilet: (0, piral_base_1.extendLoader)(loadPilet !== null && loadPilet !== void 0 ? loadPilet : (0, piral_base_1.getDefaultLoader)(loaderConfig), loaders),
        createApi,
        pilets: availablePilets,
        fetchPilets: requestPilets,
        hooks,
        dependencies: shareDependencies(modules_1.globalDependencies),
    };
    (0, app_codegen_1.integrateDebugger)(context, options, debug);
    (0, app_codegen_1.integrateEmulator)(context, options, debug);
    return options;
}
exports.createPiletOptions = createPiletOptions;
