// piral-core/lib/actions/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.includeProvider = exports.setRoute = exports.setErrorComponent = exports.setComponent = exports.injectPilet = exports.removePilet = exports.addPilet = exports.initialize = void 0;
const piral_base_1 = require("piral-base");
const utils_1 = require("../utils");
function initialize(ctx, loading, error, modules) {
    ctx.dispatch((state) => (Object.assign(Object.assign({}, state), { app: Object.assign(Object.assign({}, state.app), { error,
            loading }), modules })));
}
exports.initialize = initialize;
function addPilet(ctx, meta) {
    return ctx.options
        .loadPilet(meta)
        .then((pilet) => ctx.injectPilet(pilet))
        .then((pilet) => (0, piral_base_1.runPilet)(ctx.options.createApi, pilet, ctx.options.hooks))
        .then(utils_1.noop);
}
exports.addPilet = addPilet;
function removePilet(ctx, name) {
    ctx.dispatch((state) => (Object.assign(Object.assign({}, state), { modules: state.modules.filter((m) => m.name !== name), registry: (0, utils_1.removeNested)(state.registry, (m) => m.pilet === name) })));
    ctx.emit('unload-pilet', {
        name,
    });
    return Promise.resolve();
}
exports.removePilet = removePilet;
function injectPilet(ctx, pilet) {
    ctx.dispatch((state) => (Object.assign(Object.assign({}, state), { modules: (0, utils_1.replaceOrAddItem)(state.modules, pilet, (m) => m.name === pilet.name), registry: (0, utils_1.removeNested)(state.registry, (m) => m.pilet === pilet.name) })));
    ctx.emit('unload-pilet', {
        name: pilet.name,
    });
    return pilet;
}
exports.injectPilet = injectPilet;
function setComponent(ctx, name, component) {
    ctx.dispatch((state) => (Object.assign(Object.assign({}, state), { components: (0, utils_1.withKey)(state.components, name, component) })));
}
exports.setComponent = setComponent;
function setErrorComponent(ctx, type, component) {
    ctx.dispatch((state) => (Object.assign(Object.assign({}, state), { errorComponents: (0, utils_1.withKey)(state.errorComponents, type, component) })));
}
exports.setErrorComponent = setErrorComponent;
function setRoute(ctx, path, component) {
    ctx.dispatch((0, utils_1.withRoute)(path, component));
}
exports.setRoute = setRoute;
function includeProvider(ctx, provider) {
    ctx.dispatch((0, utils_1.withProvider)(provider));
}
exports.includeProvider = includeProvider;
