// piral-core/lib/defaults/navigator_none.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNavigation = exports.createRedirect = exports.useCurrentNavigation = exports.useRouterContext = void 0;
const _noop = () => { };
function useRouterContext() {
    return undefined;
}
exports.useRouterContext = useRouterContext;
function useCurrentNavigation() { }
exports.useCurrentNavigation = useCurrentNavigation;
function createRedirect() {
    return () => null;
}
exports.createRedirect = createRedirect;
function createNavigation() {
    return {
        get path() {
            return location.pathname;
        },
        get url() {
            const loc = location;
            return `${loc.pathname}${loc.search}${loc.hash}`;
        },
        push(target, state) { },
        replace(target, state) { },
        go(n) { },
        block(blocker) {
            return _noop;
        },
        listen(listener) {
            return _noop;
        },
        router: undefined,
    };
}
exports.createNavigation = createNavigation;
