// piral-dashboard/lib/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withRoutes = exports.withTiles = exports.getTiles = exports.getPreferences = void 0;
const default_1 = require("./default");
const Dashboard_1 = require("./Dashboard");
function getPreferences(defaultPreferences, customPreferences = {}) {
    return Object.assign(Object.assign({}, defaultPreferences), customPreferences);
}
exports.getPreferences = getPreferences;
function getTiles(items, defaultPreferences) {
    const tiles = {};
    let i = 0;
    for (const { component, preferences } of items) {
        tiles[`global-${i++}`] = {
            pilet: undefined,
            component,
            preferences: getPreferences(defaultPreferences, preferences),
        };
    }
    return tiles;
}
exports.getTiles = getTiles;
function withTiles(tiles) {
    return (state) => (Object.assign(Object.assign({}, state), { components: Object.assign({ DashboardTile: default_1.DefaultTile, DashboardContainer: default_1.DefaultContainer }, state.components), registry: Object.assign(Object.assign({}, state.registry), { tiles }) }));
}
exports.withTiles = withTiles;
function withRoutes(routes) {
    return (state) => (Object.assign(Object.assign({}, state), { routes: Object.assign(Object.assign({}, state.routes), routes.reduce((newRoutes, route) => {
            newRoutes[route] = Dashboard_1.Dashboard;
            return newRoutes;
        }, {})) }));
}
exports.withRoutes = withRoutes;
