// piral-cli/lib/common/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.legacyCoreExternals = exports.declarationEntryExtensions = exports.bundlerNames = exports.entryModuleExtensions = exports.piralJsonSchemaUrl = exports.piletJsonSchemaUrl = exports.frameworkLibs = exports.piralBaseRoot = exports.filesOnceTar = exports.packageJson = exports.filesTar = exports.defaultRegistry = void 0;
exports.defaultRegistry = 'https://registry.npmjs.org/';
exports.filesTar = 'files';
exports.packageJson = 'package.json';
exports.filesOnceTar = 'files_once';
exports.piralBaseRoot = 'piral-base/package.json';
exports.frameworkLibs = ['piral-native', 'piral', 'piral-core', 'piral-base'];
exports.piletJsonSchemaUrl = 'https://docs.piral.io/schemas/pilet-v0.json';
exports.piralJsonSchemaUrl = 'https://docs.piral.io/schemas/piral-v0.json';
exports.entryModuleExtensions = ['.ts', '.tsx', '.js', '.jsx'];
exports.bundlerNames = [
    'esbuild',
    'parcel',
    'parcel2',
    'rollup',
    'rspack',
    'webpack',
    'webpack5',
    'vite',
    'xbuild',
];
exports.declarationEntryExtensions = ['.html', '.pug', ...exports.entryModuleExtensions];
exports.legacyCoreExternals = [
    '@dbeining/react-atom',
    '@libre/atom',
    'history',
    'react',
    'react-dom',
    'react-router',
    'react-router-dom',
    'tslib',
    'path-to-regexp',
];
